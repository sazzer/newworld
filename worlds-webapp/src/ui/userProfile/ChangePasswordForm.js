// @flow

import React from 'react';
import {Button, Form} from 'semantic-ui-react'
import {Interpolate} from 'react-i18next';
import ChangePasswordFormErrorMessage from './ChangePasswordFormErrorMessage';

/** The flow type representing the props for the Change Password Form */
type ChangePasswordFormProps = {
    passwordMismatch: boolean,
    onUpdateOldPassword: (string) => void,
    onUpdateNew1Password: (string) => void,
    onUpdateNew2Password: (string) => void,
    onSave: () => void,
};

/**
 * Helper to process the onChange event from a control and call a handler with the new value
 */
function processChange(handler) {
    return function(e) {
        handler(e.target.value);
    }
}

/**
 * The form to change the users password
 */
export default function ChangePasswordForm(props: ChangePasswordFormProps) {
    let message, error;

    if (props.passwordMismatch) {
        message = <ChangePasswordFormErrorMessage errorCode='passwordMismatch' />;
        error = true;
    }

    return (
        <Form data-test="changePasswordForm" error={error}>
            <Form.Field required>
                <label><Interpolate i18nKey="userProfile.password.original" /></label>
                <input name="original"
                       type="password"
                       onChange={processChange(props.onUpdateOldPassword)}
                       required
                       autoFocus/>
            </Form.Field>
            <Form.Field required>
                <label><Interpolate i18nKey="userProfile.password.new1" /></label>
                <input name="new1"
                       type="password"
                       onChange={processChange(props.onUpdateNew1Password)}
                       required />
            </Form.Field>
            <Form.Field required>
                <label><Interpolate i18nKey="userProfile.password.new2" /></label>
                <input name="new2"
                       type="password"
                       onChange={processChange(props.onUpdateNew2Password)}
                       required />
            </Form.Field>
            { message }
            <Button primary
                    disabled={props.passwordMismatch}
                    onClick={props.onSave}><Interpolate i18nKey="userProfile.password.save" /></Button>
        </Form>
    );
}
