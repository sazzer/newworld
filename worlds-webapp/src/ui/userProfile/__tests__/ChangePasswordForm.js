// @flow

import React from 'react';
import {shallow} from 'enzyme';
import ChangePasswordForm  from '../ChangePasswordForm';

/** Create the component to test */
function setup({passwordMismatch, status, errorCode}: {passwordMismatch?: boolean, status?: string, errorCode?: string} = {}) {
    const onUpdateOldPassword = jest.fn();
    const onUpdateNew1Password = jest.fn();
    const onUpdateNew2Password = jest.fn();
    const onSave = jest.fn();

    return {
        onUpdateOldPassword,
        onUpdateNew1Password,
        onUpdateNew2Password,
        onSave,
        element: shallow(<ChangePasswordForm passwordMismatch={passwordMismatch || false}
                                             onUpdateOldPassword={onUpdateOldPassword}
                                             onUpdateNew1Password={onUpdateNew1Password}
                                             onUpdateNew2Password={onUpdateNew2Password}
                                             onSave={onSave}
                                             status={status}
                                             errorCode={errorCode} />)
    };
}

describe('ChangePasswordForm', () => {
    describe('Renders correctly', () => {
        it('When viewing a blank form', () => {
            const { element } = setup();

            expect(element).toMatchSnapshot();
        });
        it('When passwords are mismatched', () => {
            const { element } = setup({passwordMismatch: true});

            expect(element).toMatchSnapshot();
        });
        it('When successfully saved a profile', () => {
            const { element } = setup({status: 'success'});

            expect(element).toMatchSnapshot();
        });
        it('When failed to save a profile', () => {
            const { element } = setup({status: 'error', errorCode: 'oops'});

            expect(element).toMatchSnapshot();
        });
    });

    it('Triggers onSave when the save button is clicked', () => {
        const { element, onSave, onUpdateOldPassword, onUpdateNew1Password, onUpdateNew2Password } = setup();

        element.find("Button").simulate("click");

        expect(onSave).toHaveBeenCalledTimes(1);

        expect(onUpdateOldPassword).toHaveBeenCalledTimes(0);
        expect(onUpdateNew1Password).toHaveBeenCalledTimes(0);
        expect(onUpdateNew2Password).toHaveBeenCalledTimes(0);
    });

    it('Triggers onUpdateOldPassword when the old password is changed', () => {
        const { element, onSave, onUpdateOldPassword, onUpdateNew1Password, onUpdateNew2Password } = setup();

        element.find("input[name='original']").simulate('change', {
            target: {
                value: 'password'
            }
        });

        expect(onUpdateOldPassword).toHaveBeenCalledTimes(1);
        expect(onUpdateOldPassword).toHaveBeenCalledWith('password');

        expect(onUpdateNew1Password).toHaveBeenCalledTimes(0);
        expect(onUpdateNew2Password).toHaveBeenCalledTimes(0);
        expect(onSave).toHaveBeenCalledTimes(0);
    });

    it('Triggers onUpdateNew1Password when the new password is changed', () => {
        const { element, onSave, onUpdateOldPassword, onUpdateNew1Password, onUpdateNew2Password } = setup();

        element.find("input[name='new1']").simulate('change', {
            target: {
                value: 'newPassword'
            }
        });

        expect(onUpdateNew1Password).toHaveBeenCalledTimes(1);
        expect(onUpdateNew1Password).toHaveBeenCalledWith('newPassword');

        expect(onUpdateOldPassword).toHaveBeenCalledTimes(0);
        expect(onUpdateNew2Password).toHaveBeenCalledTimes(0);
        expect(onSave).toHaveBeenCalledTimes(0);
    });

    it('Triggers onUpdateNew2Password when the repeat password is changed', () => {
        const { element, onSave, onUpdateOldPassword, onUpdateNew1Password, onUpdateNew2Password } = setup();

        element.find("input[name='new2']").simulate('change', {
            target: {
                value: 'newPassword'
            }
        });

        expect(onUpdateNew2Password).toHaveBeenCalledTimes(1);
        expect(onUpdateNew2Password).toHaveBeenCalledWith('newPassword');

        expect(onUpdateOldPassword).toHaveBeenCalledTimes(0);
        expect(onUpdateNew1Password).toHaveBeenCalledTimes(0);
        expect(onSave).toHaveBeenCalledTimes(0);
    });
});
