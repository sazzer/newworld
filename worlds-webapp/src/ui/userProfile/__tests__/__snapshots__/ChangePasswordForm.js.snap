// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangePasswordForm Renders correctly When failed to save a profile 1`] = `
<Form
  as="form"
  data-test="changePasswordForm"
  error={true}
  loading={false}
  success={false}
>
  <FormField
    required={true}
  >
    <label>
      <Interpolate
        className=""
        i18nKey="userProfile.password.original"
      />
    </label>
    <input
      autoFocus={true}
      name="original"
      onChange={[Function]}
      required={true}
      type="password"
    />
  </FormField>
  <FormField
    required={true}
  >
    <label>
      <Interpolate
        className=""
        i18nKey="userProfile.password.new1"
      />
    </label>
    <input
      name="new1"
      onChange={[Function]}
      required={true}
      type="password"
    />
  </FormField>
  <FormField
    required={true}
  >
    <label>
      <Interpolate
        className=""
        i18nKey="userProfile.password.new2"
      />
    </label>
    <input
      name="new2"
      onChange={[Function]}
      required={true}
      type="password"
    />
  </FormField>
  <ChangePasswordFormErrorMessage
    errorCode="oops"
  />
  <Button
    as="button"
    onClick={[Function]}
    primary={true}
    role="button"
  >
    <Interpolate
      className=""
      i18nKey="userProfile.password.save"
    />
  </Button>
</Form>
`;

exports[`ChangePasswordForm Renders correctly When passwords are mismatched 1`] = `
<Form
  as="form"
  data-test="changePasswordForm"
  error={true}
  loading={false}
  success={false}
>
  <FormField
    required={true}
  >
    <label>
      <Interpolate
        className=""
        i18nKey="userProfile.password.original"
      />
    </label>
    <input
      autoFocus={true}
      name="original"
      onChange={[Function]}
      required={true}
      type="password"
    />
  </FormField>
  <FormField
    required={true}
  >
    <label>
      <Interpolate
        className=""
        i18nKey="userProfile.password.new1"
      />
    </label>
    <input
      name="new1"
      onChange={[Function]}
      required={true}
      type="password"
    />
  </FormField>
  <FormField
    required={true}
  >
    <label>
      <Interpolate
        className=""
        i18nKey="userProfile.password.new2"
      />
    </label>
    <input
      name="new2"
      onChange={[Function]}
      required={true}
      type="password"
    />
  </FormField>
  <ChangePasswordFormErrorMessage
    errorCode="passwordMismatch"
  />
  <Button
    as="button"
    disabled={true}
    onClick={[Function]}
    primary={true}
    role="button"
  >
    <Interpolate
      className=""
      i18nKey="userProfile.password.save"
    />
  </Button>
</Form>
`;

exports[`ChangePasswordForm Renders correctly When successfully saved a profile 1`] = `
<Form
  as="form"
  data-test="changePasswordForm"
  loading={false}
  success={true}
>
  <FormField
    required={true}
  >
    <label>
      <Interpolate
        className=""
        i18nKey="userProfile.password.original"
      />
    </label>
    <input
      autoFocus={true}
      name="original"
      onChange={[Function]}
      required={true}
      type="password"
    />
  </FormField>
  <FormField
    required={true}
  >
    <label>
      <Interpolate
        className=""
        i18nKey="userProfile.password.new1"
      />
    </label>
    <input
      name="new1"
      onChange={[Function]}
      required={true}
      type="password"
    />
  </FormField>
  <FormField
    required={true}
  >
    <label>
      <Interpolate
        className=""
        i18nKey="userProfile.password.new2"
      />
    </label>
    <input
      name="new2"
      onChange={[Function]}
      required={true}
      type="password"
    />
  </FormField>
  <ChangePasswordFormSuccessMessage />
  <Button
    as="button"
    onClick={[Function]}
    primary={true}
    role="button"
  >
    <Interpolate
      className=""
      i18nKey="userProfile.password.save"
    />
  </Button>
</Form>
`;

exports[`ChangePasswordForm Renders correctly When viewing a blank form 1`] = `
<Form
  as="form"
  data-test="changePasswordForm"
  loading={false}
  success={false}
>
  <FormField
    required={true}
  >
    <label>
      <Interpolate
        className=""
        i18nKey="userProfile.password.original"
      />
    </label>
    <input
      autoFocus={true}
      name="original"
      onChange={[Function]}
      required={true}
      type="password"
    />
  </FormField>
  <FormField
    required={true}
  >
    <label>
      <Interpolate
        className=""
        i18nKey="userProfile.password.new1"
      />
    </label>
    <input
      name="new1"
      onChange={[Function]}
      required={true}
      type="password"
    />
  </FormField>
  <FormField
    required={true}
  >
    <label>
      <Interpolate
        className=""
        i18nKey="userProfile.password.new2"
      />
    </label>
    <input
      name="new2"
      onChange={[Function]}
      required={true}
      type="password"
    />
  </FormField>
  <Button
    as="button"
    disabled={false}
    onClick={[Function]}
    primary={true}
    role="button"
  >
    <Interpolate
      className=""
      i18nKey="userProfile.password.save"
    />
  </Button>
</Form>
`;
